### PR TITLE
encryption: fix freezing key controller

### PR DESCRIPTION
**Pull request by sttts**
The key controller was still just expecting one read-key, while with the recent change for backup/restore we always preserve one more key. This made the key controller freeze after the first two read-keys were in the config.

While fixing the upper, we also run into the case that during pruning this one more read-key is preserved, but this read-key must be backed. Hence, we have to preserve also all other unbacked read-keys in between.

### CODE REVIEW

File: pkg/operator/encryption/controllers/key_controller_test.go [hunk 1]
```diff
@@ -210,6 +212,48 @@ func TestKeyController(t *testing.T) {
 			},
 		},
 
+		{
+			name: "create a new write key when the previous key expired and another read key exists",
```
Comment (p0lyn0mial) [inline pkg/operator/encryption/controllers/key_controller_test.go]: shouldn't we validate if a new key was created? (`validateFunc` function)

File: pkg/operator/encryption/controllers/key_controller_test.go [hunk 2]
```diff
@@ -210,6 +212,48 @@ func TestKeyController(t *testing.T) {
...
+					return ecs
+				}(),
+			},
+			apiServerObjects: apiServerAesCBC,
+			targetNamespace:  "kms",
+			expectedActions: []string{
+				"list:pods:kms",
+				"get:secrets:kms",
+				"list:secrets:openshift-config-managed",
+				"create:secrets:openshift-config-managed",
```
Comment (p0lyn0mial) [inline pkg/operator/encryption/controllers/key_controller_test.go]: this only says that something was created ;)

File: pkg/operator/encryption/state/helpers.go [hunk 3]
```diff
@@ -34,11 +34,15 @@ func MigratedFor(grs []schema.GroupResource, km KeyState) (ok bool, missing []sc
...
 }
 
-// KeysWithPotentiallyPersistedData returns the minimal, recent secrets which have migrated all given GRs.
-func KeysWithPotentiallyPersistedData(grs []schema.GroupResource, recentFirstSortedKeys []KeyState) []KeyState {
+// KeysWithPotentiallyPersistedDataAndNextReadKey returns the minimal, recent secrets which have migrated all given GRs.
+func KeysWithPotentiallyPersistedDataAndNextReadKey(grs []schema.GroupResource, recentFirstSortedKeys []KeyState) []KeyState {
 	for i, k := range recentFirstSortedKeys {
 		if allMigrated, missing, _ := MigratedFor(grs, k); allMigrated {
-			return recentFirstSortedKeys[:i+1]
+			if i+1 < len(recentFirstSortedKeys) {
```
Comment (p0lyn0mial) [inline pkg/operator/encryption/state/helpers.go]: this looks to be safe, we could have a unit test for it.

File: pkg/operator/encryption/controllers/key_controller_test.go (hunk 2 above)
Comment (sttts) [reply pkg/operator/encryption/controllers/key_controller_test.go]: that's a general issue in this test, isn't it? But what else do we create in the key controller than keys? :)

File: pkg/operator/encryption/state/helpers.go (hunk 3 above)
Comment (sttts) [reply pkg/operator/encryption/state/helpers.go]: what do you want to test?

Comment (sttts) [reply pkg/operator/encryption/state/helpers.go]: we have tests that "one more read-key" is fine. So this is covered. The rest here is just Golang standard behaviour.

Comment (p0lyn0mial) [reply pkg/operator/encryption/state/helpers.go]: sgtm.

File: pkg/operator/encryption/controllers/key_controller_test.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/operator/encryption/controllers/key_controller_test.go]: yeah, hard to beat that argument :)

**Pull request merged**